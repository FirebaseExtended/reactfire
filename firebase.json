{
  "emulators": {
    "auth": {
      "port": 9099
    },
    "functions": {
      "port": 5001
    },
    "firestore": {
      "port": 8080
    },
    "database": {
      "port": 9000
    },
    "storage": {
      "port": 9199
    },
    "ui": {
      "enabled": true
    }
  },
  "storage": {
    "rules": "storage.rules"
  }
}
