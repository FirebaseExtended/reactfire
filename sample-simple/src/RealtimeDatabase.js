import React, { useState } from 'react';
import {
  AuthCheck,
  SuspenseWithPerf,
  useDatabaseList,
  useDatabaseObject,
  useDatabase
} from 'reactfire';

const Counter = props => {
  const database = useDatabase();
  const ref = database().ref('counter');
  const increment = amountToIncrement => {
    ref.transaction(counterVal => {
      return counterVal + amountToIncrement;
    });
  };

  const { snapshot } = useDatabaseObject(ref);
  const counterValue = snapshot.val();

  return (
    <>
      <button onClick={() => increment(-1)}>-</button>
      <span> {counterValue} </span>
      <button onClick={() => increment(1)}>+</button>
    </>
  );
};

const AnimalEntry = ({ saveAnimal }) => {
  const [text, setText] = useState('');
  const [disabled, setDisabled] = useState(false);

  const onSave = () => {
    setDisabled(true);
    saveAnimal(text).then(() => {
      setText('');
      setDisabled(false);
    });
  };

  return (
    <>
      <input
        value={text}
        disabled={disabled}
        placeholder="Iguana"
        onChange={({ target }) => setText(target.value)}
      />
      <button onClick={onSave} disabled={disabled || text.length < 3}>
        Add new animal
      </button>
    </>
  );
};

const List = props => {
  const database = useDatabase();
  const ref = database().ref('animals');
  const changes = useDatabaseList(ref);
  const addNewAnimal = commonName => {
    const newAnimalRef = ref.push();
    return newAnimalRef.set({
      commonName
    });
  };

  const removeAnimal = id => ref.child(id).remove();

  return (
    <>
      <AnimalEntry saveAnimal={addNewAnimal} />
      <ul>
        {changes.map(({ snapshot }) => (
          <li key={snapshot.key}>
            {snapshot.val().commonName}{' '}
            <button onClick={() => removeAnimal(snapshot.key)}>X</button>
          </li>
        ))}
      </ul>
    </>
  );
};

const SuspenseWrapper = props => {
  return (
    <SuspenseWithPerf fallback="loading..." traceId="RTDB-root">
      <AuthCheck fallback="sign in to use Realtime Database">
        <h3>Sample Object Listener</h3>
        <Counter />
        <h3>Sample List Listener</h3>
        <List />
      </AuthCheck>
    </SuspenseWithPerf>
  );
};

export default SuspenseWrapper;
