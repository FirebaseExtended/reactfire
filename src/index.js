import createAppContainer from './createAppContainer';
import createAuthContainer from './createAuthContainer';
// import createDatabaseContainer from './createDatabaseContainer';

export {
  createAppContainer,
  createAuthContainer,
  // createDatabaseContainer,
};
