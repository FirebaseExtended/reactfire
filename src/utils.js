/**
 * Returns whether or not the provided input is a JavaScript function.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a JavaScript function.
 */
export function isFunction(input) {
  return typeof input === 'function';
}

/**
 * Returns whether or not the provided input is a non-null JavaScript object.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a non-null JavaScript object.
 */
export function isNonNullObject(input) {
  return Object.prototype.toString.call(input) === '[object Object]';
}

/**
 * Returns whether or not the provided input is a Firebase Database Reference.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a Firebase Database Reference.
 */
export function isDatabaseReference(input) {
  return isNonNullObject(input) &&
    typeof input.ref === 'object' &&
    typeof input.on === 'function' &&
    typeof input.set === 'function' &&
    typeof input.once === 'function' &&
    typeof input.transaction === 'function';
}

/**
 * Returns whether or not the provided input is a Firebase Database Query.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a Firebase Database Query.
 */
export function isDatabaseQuery(input) {
  return isNonNullObject(input) &&
    typeof input.ref === 'object' &&
    typeof input.on === 'function' &&
    typeof input.once === 'function' &&
    typeof input.endAt === 'function' &&
    typeof input.orderByChild === 'function';
}

/**
 * Returns whether or not the provided input is either a Firebase Database Reference or a Firebase
 * Database Query.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is either a Firebase Database Reference or
 * a Firebase Database Query.
 */
export function isDatabaseReferenceOrQuery(input) {
  return isDatabaseReference(input) || isDatabaseQuery(input);
}

/**
 * Returns whether or not the provided input is a Firebase App.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a Firebase App.
 */
export function isFirebaseApp(input) {
  return isNonNullObject(input) &&
    typeof input.name === 'string' &&
    typeof input.options === 'object' &&
    typeof input.delete === 'function';
}

/**
 * Returns whether or not the provided input is a React component.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a React component.
 */
export function isReactComponent(input) {
  return isNonNullObject(input) || isFunction(input);
}

/**
 * Returns the display name for the provided component, used by things like React Dev Tools.
 *
 * @param {React.Component} Component The component whose display name to return.
 *
 * @return {string} The provide component's display name.
 */
export function getDisplayName(Component) {
  return Component.displayName || Component.name || 'Component';
}

// /**
//  * Returns the index of the key in the list. If an item with the key is not in the list, -1 is
//  * returned.
//  *
//  * @param {Array<any>} list A list of items.
//  * @param {string} key The key for which to search.
//  *
//  * @return {number} The index of the item which has the provided key or -1 if no items have the
//  * provided key.
//  */
// export function indexForKey(list, key) {
//   for (let i = 0, length = list.length; i < length; i++) {  // eslint-disable-line no-plusplus
//     if (list[i]['.key'] === key) {
//       return i;
//     }
//   }

//   /* istanbul ignore next */
//   return -1;
// }

// /**
//  * Creates a new record given a key-value pair.
//  *
//  * @param {string} key The new record's key.
//  * @param {any} value The new record's value.
//  *
//  * @return {Object} The new record.
//  */
// export function createRecord(key, value) {
//   let record = {};
//   if (typeof value === 'object' && value !== null) {
//     record = value;
//   } else {
//     record['.value'] = value;
//   }
//   record['.key'] = key;

//   return record;
// }
