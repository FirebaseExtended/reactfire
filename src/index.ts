import firebase from 'firebase/app';
import { SuspenseSubject } from './SuspenseSubject';

export type ReactFireGlobals = {
  _reactFireDatabaseCachedQueries: Array<firebase.database.Query>;
  _reactFireFirestoreQueryCache: Array<firebase.firestore.Query>;
  _reactFirePreloadedObservables: Map<string, SuspenseSubject<any>>;
};

export interface ReactFireOptions<T = unknown> {
  idField?: string;
  initialData?: T | any;
  suspense?: boolean;
}

export function checkOptions(options: ReactFireOptions, field: string) {
  // make sure the field passed in is a valid key of ReactFire Options
  if (field === 'idField' || field === 'initialData' || field === 'suspense') {
    return options ? (options[field] as ReactFireOptions['idField'] | ReactFireOptions['initialData'] | ReactFireOptions['suspense']) : undefined;
  }

  throw new Error(`Field "${field}" is not a valid key in ReactFireOptions`);
}

export function checkinitialData(options: ReactFireOptions) {
  return checkOptions(options, 'initialData');
}

export function checkIdField(options: ReactFireOptions) {
  return checkOptions(options, 'idField');
}

export * from './auth';
export * from './database';
export * from './firebaseApp';
export * from './firestore';
export * from './performance';
export * from './remote-config';
export * from './storage';
export * from './useObservable';
