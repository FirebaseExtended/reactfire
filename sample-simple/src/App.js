import React from 'react';
import AuthButton from './Auth';
import FirestoreCounter from './Firestore';
import Storage from './Storage';
import RealtimeDatabase from './RealtimeDatabase';
import {
  preloadFirestoreDoc,
  useFirebaseApp,
  preloadUser,
  preloadAuth,
  preloadFirestore,
  preloadDatabase,
  preloadStorage
} from 'reactfire';

const Fire = () => (
  <span role="img" aria-label="Fire">
    ðŸ”¥
  </span>
);

const Card = ({ title, children }) => {
  return (
    <div className="card">
      <h1>
        {title} <Fire />
      </h1>
      {children}
    </div>
  );
};

const preloadSDKs = firebaseApp => {
  return Promise.all([
    preloadFirestore(firebaseApp),
    preloadDatabase(firebaseApp),
    preloadStorage(firebaseApp),
    preloadAuth(firebaseApp)
  ]);
};

const preloadData = async firebaseApp => {
  const user = await preloadUser(firebaseApp);

  if (user) {
    preloadFirestoreDoc(
      firestore => firestore.doc('count/counter'),
      firebaseApp
    );
  }
};

const App = () => {
  const firebaseApp = useFirebaseApp();

  // kick off stuff we know our components will eventually need
  preloadSDKs(firebaseApp).then(preloadData(firebaseApp));

  return (
    <>
      <h1>
        <Fire /> ReactFire Demo <Fire />
      </h1>
      <div className="container">
        <Card title="Authentication">
          <AuthButton />
        </Card>

        <Card title="Firestore">
          <FirestoreCounter />
        </Card>

        <Card title="Storage">
          <Storage />
        </Card>

        <Card title="Realtime Database">
          <RealtimeDatabase />
        </Card>
      </div>
    </>
  );
};

export default App;
