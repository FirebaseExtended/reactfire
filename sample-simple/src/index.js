import React, { StrictMode } from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import * as serviceWorker from './serviceWorker';
import './app.css';
import { FirebaseAppProvider } from 'reactfire';
import 'firebase/performance';

const config = {
  apiKey: 'AIzaSyBg3u1sJlyJwQCE95oSDH_mtLABS-is8ZM',
  authDomain: 'rxfire-525a3.firebaseapp.com',
  databaseURL: 'https://rxfire-525a3.firebaseio.com',
  projectId: 'rxfire-525a3',
  storageBucket: 'rxfire-525a3.appspot.com',
  messagingSenderId: '844180061847',
  appId: '1:844180061847:web:400f7142e2d1aaeb'
};

ReactDOM.render(
  <StrictMode>
    <FirebaseAppProvider firebaseConfig={config} initPerformance>
      <App />
    </FirebaseAppProvider>
  </StrictMode>,
  document.getElementById('root')
); // https://reactjs.org/docs/strict-mode.html

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();
