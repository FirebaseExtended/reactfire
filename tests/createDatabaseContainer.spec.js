import React from 'react';
import renderer from 'react-test-renderer';
import * as firebase from 'firebase';

import { createDatabaseContainer } from '../src/';

import { DummyComponent, mockFirebaseAppOptions } from './utils';


describe.skip('createDatabaseContainer()', () => {
  let firebaseRef;

  beforeAll(() => firebase.initializeApp(mockFirebaseAppOptions));

  afterAll(() => firebase.app().delete());

  beforeEach(() => {
    firebaseRef = firebase.database().ref('reactfire').push();
  });

  afterEach(() => firebaseRef.remove());


  describe('Input validation', () => {
    test('should throw given nothing as the first argument', () => {
      expect(() => {
        createDatabaseContainer();
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given an invalid React component as the first argument', () => {
      expect(() => {
        createDatabaseContainer(null);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given nothing as the second argument', () => {
      expect(() => {
        createDatabaseContainer(DummyComponent);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given a non-function as the second argument', () => {
      expect(() => {
        createDatabaseContainer(DummyComponent, true);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should not throw given a function which generates an empty object as the second argument', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({}));

        renderer.create(<WrappedDummyComponent />);
      }).not.toThrow();
    });

    test('should throw given a function which generates an object containing a key whose value is a non-object', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
          item0: true,
        }));

        renderer.create(<WrappedDummyComponent />);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given a function which generates an object containing a key whose value is an object which is not a Database Reference', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
          item0: {
            foo: 'bar',
          },
        }));

        renderer.create(<WrappedDummyComponent />);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given a function which generates an object containing a key whose value is an object whose "ref" key is a non-object', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
          item0: {
            ref: true,
          },
        }));

        renderer.create(<WrappedDummyComponent />);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given a function which generates an object containing a key whose value is an object whose "ref" key is an object which is not a Database Reference', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
          item0: {
            ref: {
              foo: 'bar',
            },
          },
        }));

        renderer.create(<WrappedDummyComponent />);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given an object which contains an invalid key', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
          item0: {
            ref: firebaseRef,
            foo: true,
          },
        }));

        renderer.create(<WrappedDummyComponent />);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given an object which contains no "ref" key', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
          item0: {
            asArray: true,
          },
        }));

        renderer.create(<WrappedDummyComponent />);
      }).toThrowErrorMatchingSnapshot();
    });

    test('should throw given an object which contains a non-boolean "asArray" key', () => {
      expect(() => {
        const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
          item0: {
            ref: firebaseRef,
            asArray: 'foo',
          },
        }));

        renderer.create(<WrappedDummyComponent />);
      }).toThrowErrorMatchingSnapshot();
    });
  });


  describe('Array', () => {
    test('binds array records which are objects', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef,
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({
        first: { index: 0 },
        second: { index: 1 },
        third: { index: 2 },
      })
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'first', index: 0 },
            { '.key': 'second', index: 1 },
            { '.key': 'third', index: 2 },
          ]);
        });
    });

    test('binds array records which are primitives', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef,
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set(['first', 'second', 'third'])
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': '0', '.value': 'first' },
            { '.key': '1', '.value': 'second' },
            { '.key': '2', '.value': 'third' },
          ]);
        });
    });

    test('binds array records which are a mix of objects and primitives', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef,
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({
        0: true,
        1: 'second',
        third: { index: 2 },
      })
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': '0', '.value': true },
            { '.key': '1', '.value': 'second' },
            { '.key': 'third', index: 2 },
          ]);
        });
    });

    test('binds as an empty array for Firebase references with no data', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef,
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set(null)
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([]);
        });
    });

    test('binds sparse arrays', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef,
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ 0: 'a', 2: 'b', 5: 'c' })
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': '0', '.value': 'a' },
            { '.key': '2', '.value': 'b' },
            { '.key': '5', '.value': 'c' },
          ]);
        });
    });

    test('binds only a subset of records when using limit queries', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.limitToLast(2),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 1, b: 2, c: 3 })
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'b', '.value': 2 },
            { '.key': 'c', '.value': 3 },
          ]);
        });
    });

    test('removes records when they fall outside of a limit query', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.limitToLast(2),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 1, b: 2, c: 3 })
        .then(() => firebaseRef.child('d').set(4))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'c', '.value': 3 },
            { '.key': 'd', '.value': 4 },
          ]);
        });
    });

    test('adds a new record when an existing record in the limit query is removed', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.limitToLast(2),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 1, b: 2, c: 3 })
        .then(() => firebaseRef.child('b').remove())
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', '.value': 1 },
            { '.key': 'c', '.value': 3 },
          ]);
        });
    });

    test('binds records in the correct order when using ordered queries', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByValue(),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 2, b: 1, c: 3 })
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'b', '.value': 1 },
            { '.key': 'a', '.value': 2 },
            { '.key': 'c', '.value': 3 },
          ]);
        });
    });

    test('binds multiple Firebase references to props at the same time', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items0: {
          ref: firebaseRef.child('items0'),
          asArray: true,
        },
        items1: {
          ref: firebaseRef.child('items1'),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({
        items0: {
          first: { index: 0 },
          second: { index: 1 },
          third: { index: 2 },
        },
        items1: ['first', 'second', 'third'],
      })
        .then(() => {
          const props = component.toJSON().props;
          expect(props.items0).toEqual([
            { '.key': 'first', index: 0 },
            { '.key': 'second', index: 1 },
            { '.key': 'third', index: 2 },
          ]);
          expect(props.items1).toEqual([
            { '.key': '0', '.value': 'first' },
            { '.key': '1', '.value': 'second' },
            { '.key': '2', '.value': 'third' },
          ]);
        });
    });

    test('updates an array record when its value changes', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef,
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 1, b: 2, c: 3 })
        .then(() => firebaseRef.child('b').set({ foo: 'bar' }))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', '.value': 1 },
            { '.key': 'b', foo: 'bar' },
            { '.key': 'c', '.value': 3 },
          ]);
        });
    });

    test('removes an array record when it is deleted', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef,
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 1, b: 2, c: 3 })
        .then(() => firebaseRef.child('b').remove())
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', '.value': 1 },
            { '.key': 'c', '.value': 3 },
          ]);
        });
    });

    test('moves an array record when it\'s order changes (moved to start of array) [orderByValue()]', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByValue(),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 2, b: 3, c: 2 })
        .then(() => firebaseRef.child('b').set(1))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'b', '.value': 1 },
            { '.key': 'a', '.value': 2 },
            { '.key': 'c', '.value': 2 },
          ]);
        });
    });

    test('moves an array record when it\'s order changes (moved to middle of array) [orderByValue()]', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByValue(),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 2, b: 1, c: 4 })
        .then(() => firebaseRef.child('b').set(3))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', '.value': 2 },
            { '.key': 'b', '.value': 3 },
            { '.key': 'c', '.value': 4 },
          ]);
        });
    });

    test('moves an array record when it\'s order changes (moved to end of array) [orderByValue()]', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByValue(),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: 2, b: 1, c: 3 })
        .then(() => firebaseRef.child('b').set(4))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', '.value': 2 },
            { '.key': 'c', '.value': 3 },
            { '.key': 'b', '.value': 4 },
          ]);
        });
    });

    test('moves an array record when it\'s order changes (moved to start of array) [orderByChild()]', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByChild('value'),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: { value: 2 }, b: { value: 3 }, c: { value: 2 } })
        .then(() => firebaseRef.child('b').set({ value: 1 }))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'b', value: 1 },
            { '.key': 'a', value: 2 },
            { '.key': 'c', value: 2 },
          ]);
        });
    });

    test('moves an array record when it\'s order changes (moved to middle of array) [orderByChild()]', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByChild('value'),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: { value: 2 }, b: { value: 1 }, c: { value: 4 } })
        .then(() => firebaseRef.child('b').set({ value: 3 }))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', value: 2 },
            { '.key': 'b', value: 3 },
            { '.key': 'c', value: 4 },
          ]);
        });
    });

    test('moves an array record when it\'s order changes (moved to end of array) [orderByChild()]', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByChild('value'),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ a: { value: 2 }, b: { value: 1 }, c: { value: 3 } })
        .then(() => firebaseRef.child('b').set({ value: 4 }))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', value: 2 },
            { '.key': 'c', value: 3 },
            { '.key': 'b', value: 4 },
          ]);
        });
    });

    test('works with orderByKey() queries', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: {
          ref: firebaseRef.orderByKey(),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.set({ b: 2, c: 1, d: 3 })
        .then(() => firebaseRef.update({ a: 4, d: 4, e: 0 }))
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual([
            { '.key': 'a', '.value': 4 },
            { '.key': 'b', '.value': 2 },
            { '.key': 'c', '.value': 1 },
            { '.key': 'd', '.value': 4 },
            { '.key': 'e', '.value': 0 },
          ]);
        });
    });

    test.skip('should stop listening for changes when the component is unmounted', () => {
      // TODO
    });
  });


  describe('Object', () => {
    test('binds to an object', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: firebaseRef.child('items'),
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      const obj = {
        first: { index: 0 },
        second: { index: 1 },
        third: { index: 2 },
      };

      return firebaseRef.child('items').set(obj)
        .then(() => {
          obj['.key'] = 'items';

          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual(obj);
        });
    });

    test('binds to a primitive', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: firebaseRef.child('items'),
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.child('items').set('foo')
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual({
            '.key': 'items',
            '.value': 'foo',
          });
        });
    });

    test('binds to Firebase references with no data', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: firebaseRef.child('items'),
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.child('items').set(null)
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual({
            '.key': 'items',
            '.value': null,
          });
        });
    });

    test('sets the key as null when bound to the root of the database', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: firebaseRef.root,
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.root.once('value')
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp['.key']).toBeNull();
        });
    });

    test('binds with limit queries', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items: firebaseRef.child('items').limitToLast(2),
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      return firebaseRef.child('items').set({
        first: { index: 0 },
        second: { index: 1 },
        third: { index: 2 },
      })
        .then(() => {
          const itemsProp = component.toJSON().props.items;
          expect(itemsProp).toEqual({
            '.key': 'items',
            second: { index: 1 },
            third: { index: 2 },
          });
        });
    });

    test('binds multiple Firebase references to props at the same time', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items0: firebaseRef.child('items0'),
        items1: firebaseRef.child('items1'),
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      const items0 = {
        first: { index: 0 },
        second: { index: 1 },
        third: { index: 2 },
      };

      const items1 = {
        bar: {
          foo: 'baz',
        },
        baz: true,
        foo: 100,
      };

      return firebaseRef.set({ items0, items1 })
        .then(() => {
          items0['.key'] = 'items0';
          items1['.key'] = 'items1';

          const props = component.toJSON().props;
          expect(props.items0).toEqual(items0);
          expect(props.items1).toEqual(items1);
        });
    });

    test('binds a mixture of arrays and objects to state variables at the same time', () => {
      const WrappedDummyComponent = createDatabaseContainer(DummyComponent, () => ({
        items0: firebaseRef.child('items0'),
        items1: {
          ref: firebaseRef.child('items1'),
          asArray: true,
        },
      }));

      const component = renderer.create(<WrappedDummyComponent />);

      const items0 = {
        first: { index: 0 },
        second: { index: 1 },
        third: { index: 2 },
      };

      const items1 = {
        bar: {
          foo: 'baz',
        },
        baz: true,
        foo: 100,
      };

      return firebaseRef.set({ items0, items1 })
        .then(() => {
          items0['.key'] = 'items0';
          items1['.key'] = 'items1';

          const props = component.toJSON().props;
          expect(props.items0).toEqual(items0);
          expect(props.items1).toEqual([
            { '.key': 'bar', foo: 'baz' },
            { '.key': 'baz', '.value': true },
            { '.key': 'foo', '.value': 100 },
          ]);
        });
    });

    test.skip('should stop listening for changes when the component is unmounted', () => {
      // TODO
    });
  });
});
