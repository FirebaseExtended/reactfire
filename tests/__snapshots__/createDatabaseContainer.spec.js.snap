exports[`createDatabaseContainer() Input validation should throw given a function which generates an object containing a key whose value is a non-object 1`] = `"Unexpected reference object. Not an object."`;

exports[`createDatabaseContainer() Input validation should throw given a function which generates an object containing a key whose value is an object which is not a Database Reference 1`] = `"Reference must be a Database Reference or Query or an object with a \"ref\" key."`;

exports[`createDatabaseContainer() Input validation should throw given a function which generates an object containing a key whose value is an object whose "ref" key is a non-object 1`] = `"\"ref\" key must be a Database Reference or Query."`;

exports[`createDatabaseContainer() Input validation should throw given a function which generates an object containing a key whose value is an object whose "ref" key is an object which is not a Database Reference 1`] = `"\"ref\" key must be a Database Reference or Query."`;

exports[`createDatabaseContainer() Input validation should throw given a non-function as the second argument 1`] = `"Second argument to createDatabaseContainer() must be a function."`;

exports[`createDatabaseContainer() Input validation should throw given an invalid React component as the first argument 1`] = `"First argument to createDatabaseContainer() must be a React component."`;

exports[`createDatabaseContainer() Input validation should throw given an object which contains a non-boolean "asArray" key 1`] = `"\"asArray\" key must be a boolean."`;

exports[`createDatabaseContainer() Input validation should throw given an object which contains an invalid key 1`] = `"Unexpected key in ref: \"foo\"."`;

exports[`createDatabaseContainer() Input validation should throw given an object which contains no "ref" key 1`] = `"Reference must be a Database Reference or Query or an object with a \"ref\" key."`;

exports[`createDatabaseContainer() Input validation should throw given nothing as the first argument 1`] = `"First argument to createDatabaseContainer() must be a React component."`;

exports[`createDatabaseContainer() Input validation should throw given nothing as the second argument 1`] = `"Second argument to createDatabaseContainer() must be a function."`;
